src/poetry/core/version/markers.py

```diff
@@ -246,7 +249,7 @@ def constraint_string(self) -> str:
         return self._constraint_string
 
     @property
-    def constraint(self) -> "VersionTypes":
+    def constraint(self) -> Union["BaseConstraint", VersionConstraint]:
```

Comment:
Just curious what the reason for migrating away form the `VersionTypes` type is -- is it merely a simple, useless typedef? In that case, should we factor it out entirely in the future?

Reply:
I tend to say "yes". `VersionTypes` contains every subclass of `VersionConstraint`, but not `VersionConstraint` itself. Everywhere `VersionTypes` is used, you can only be sure to use the interface of `VersionConstraint`, otherwise an isinstance check for a subclass is required.

As for this case: At first, I only added `BaseConstraint` because it can be both (`BaseConstraint` and `VersionTypes`). Due to the fact it can be both, in `MarkerUnion.of()` I had to add `isinstance(union, VersionConstraint)` before calling a method specific to `VersionConstraint`/`VersionTypes`, afterwards calling `SingleMarker()` with this union. Then, mypy tells me that `BaseConstraint` or `VersionTypes` is required but `VersionConstraint` is given...

Reply:
The future ticket has overtaken this one. 😉 (VersionTypes has been removed in #302)